File: backend/src/test/java/org/globe42/web/charges/ChargeTypeControllerMvcTest.java
package org.globe42.web.charges;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.ChargeCategoryDao;
import org.globe42.dao.ChargeTypeDao;
import org.globe42.domain.ChargeCategory;
import org.globe42.domain.ChargeType;
import org.globe42.test.GlobeMvcTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC tests for {@link ChargeTypeController}
 * @author JB Nizet
 */
@RunWith(SpringRunner.class)
@GlobeMvcTest(ChargeTypeController.class)
public class ChargeTypeControllerMvcTest {

    @MockBean
    private ChargeTypeDao mockChargeTypeDao;

    @MockBean
    private ChargeCategoryDao mockChargeCategoryDao;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private MockMvc mvc;

    private ChargeType chargeType;

    @Before
    public void prepare() {
        chargeType = new ChargeType(42L);
        chargeType.setName("source 1");
        chargeType.setCategory(new ChargeCategory(1L, "category 1"));
        chargeType.setMaxMonthlyAmount(new BigDecimal("1234.56"));
    }

    @Test
    public void shouldList() throws Exception {
        when(mockChargeTypeDao.findAll()).thenReturn(Collections.singletonList(chargeType));

        mvc.perform(get("/api/charge-types"))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(42))
           .andExpect(jsonPath("$[0].name").value(chargeType.getName()))
           .andExpect(jsonPath("$[0].category.id").value(chargeType.getCategory().getId()))
           .andExpect(jsonPath("$[0].category.name").value(chargeType.getCategory().getName()))
           .andExpect(jsonPath("$[0].maxMonthlyAmount").value(chargeType.getMaxMonthlyAmount().doubleValue()));
    }

    @Test
    public void shouldGet() throws Exception {
        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));

        mvc.perform(get("/api/charge-types/{typeId}", chargeType.getId()))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$.id").value(42));
    }

    @Test
    public void shouldCreate() throws Exception {
        ChargeTypeCommandDTO command = ChargeTypeControllerTest.createCommand();
        when(mockChargeTypeDao.save(any(ChargeType.class))).thenReturn(chargeType);
        when(mockChargeCategoryDao.findById(command.getCategoryId())).thenReturn(Optional.of(chargeType.getCategory()));

        mvc.perform(post("/api/charge-types")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.id").value(42));
    }

    @Test
    public void shouldUpdate() throws Exception {
        ChargeTypeCommandDTO command = ChargeTypeControllerTest.createCommand();

        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));
        when(mockChargeCategoryDao.findById(command.getCategoryId())).thenReturn(Optional.of(chargeType.getCategory()));

        mvc.perform(put("/api/charge-types/{sourceId}", chargeType.getId())
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isNoContent());
    }
}

File: backend/src/test/kotlin/org/globe42/web/charges/ChargeTypeControllerMvcTest.kt
package org.globe42.web.charges

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.globe42.dao.ChargeCategoryDao
import org.globe42.dao.ChargeTypeDao
import org.globe42.domain.ChargeCategory
import org.globe42.domain.ChargeType
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import org.springframework.test.web.servlet.put
import java.math.BigDecimal

/**
 * MVC tests for [ChargeTypeController]
 * @author JB Nizet
 */
@GlobeMvcTest(ChargeTypeController::class)
class ChargeTypeControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {

    @MockkBean
    private lateinit var mockChargeTypeDao: ChargeTypeDao

    @MockkBean
    private lateinit var mockChargeCategoryDao: ChargeCategoryDao

    private lateinit var chargeType: ChargeType

    @BeforeEach
    fun prepare() {
        chargeType = ChargeType(42L).apply {
            name = "source 1"
            category = ChargeCategory(1L, "category 1")
            maxMonthlyAmount = BigDecimal("1234.56")
        }
    }

    @Test
    fun `should list`() {
        every { mockChargeTypeDao.findAll() }  returns listOf(chargeType)

        mvc.get("/api/charge-types").andExpect {
            status { isOk }
            jsonValue("$[0].id", 42)
            jsonValue("$[0].name", chargeType.name)
            jsonValue("$[0].category.id", chargeType.category.id!!)
            jsonValue("$[0].category.name", chargeType.category.name)
            jsonValue("$[0].maxMonthlyAmount", chargeType.maxMonthlyAmount!!.toDouble())
        }
    }

    @Test
    fun `should get`() {
        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType

        mvc.get("/api/charge-types/{typeId}", chargeType.id).andExpect {
            status { isOk }
            jsonValue("$.id", 42)
        }
    }

    @Test
    fun `should create`() {
        val command = createCommand()
        every { mockChargeTypeDao.existsByName(command.name) } returns false
        every { mockChargeTypeDao.save(any<ChargeType>())} returns chargeType
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns chargeType.category

        mvc.post("/api/charge-types") {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isCreated }
            jsonValue("$.id", 42)
        }
    }

    @Test
    fun `should update`() {
        val command = createCommand()

        every { mockChargeTypeDao.findByName(command.name) } returns null
        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns chargeType.category

        mvc.put("/api/charge-types/{sourceId}", chargeType.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isNoContent }
        }
    }
}
